# Homework  
Data documentation: https://tidyr.tidyverse.org/reference/who.html

__Directions:__ Answer the three questions below. Create as many code/markdown blocks necessary to answer the questions. 

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

df = pd.read_csv('https://raw.githubusercontent.com/msaricaumbc/DS_data/master/ds602/who.csv')

In [4]:
df.head()

country iso2 iso3  year  new_sp_m014  new_sp_m1524  new_sp_m2534  \
0  Afghanistan   AF  AFG  1980          NaN           NaN           NaN   
1  Afghanistan   AF  AFG  1981          NaN           NaN           NaN   
2  Afghanistan   AF  AFG  1982          NaN           NaN           NaN   
3  Afghanistan   AF  AFG  1983          NaN           NaN           NaN   
4  Afghanistan   AF  AFG  1984          NaN           NaN           NaN   

   new_sp_m3544  new_sp_m4554  new_sp_m5564  ...  newrel_m4554  newrel_m5564  \
0           NaN           NaN           NaN  ...           NaN           NaN   
1           NaN           NaN           NaN  ...           NaN           NaN   
2           NaN           NaN           NaN  ...           NaN           NaN   
3           NaN           NaN           NaN  ...           NaN           NaN   
4           NaN           NaN           NaN  ...           NaN           NaN   

   newrel_m65  newrel_f014  newrel_f1524  newrel_f2534  newrel_f3544  \
0         NaN          NaN           NaN           NaN           NaN   
1         NaN          NaN           NaN           NaN           NaN   
2         NaN          NaN           NaN           NaN           NaN   
3         NaN          NaN           NaN           NaN           NaN   
4         NaN          NaN           NaN           NaN           NaN   

   newrel_f4554  newrel_f5564  newrel_f65  
0           NaN           NaN         NaN  
1           NaN           NaN         NaN  
2           NaN           NaN         NaN  
3           NaN           NaN         NaN  
4           NaN           NaN         NaN  

[5 rows x 60 columns]

In [39]:
df = df.drop(columns=['iso2', 'iso3'])
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7240 entries, 0 to 7239
Data columns (total 58 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   country       7240 non-null   object 
 1   year          7240 non-null   int64  
 2   new_sp_m014   3173 non-null   float64
 3   new_sp_m1524  3209 non-null   float64
 4   new_sp_m2534  3206 non-null   float64
 5   new_sp_m3544  3219 non-null   float64
 6   new_sp_m4554  3223 non-null   float64
 7   new_sp_m5564  3218 non-null   float64
 8   new_sp_m65    3209 non-null   float64
 9   new_sp_f014   3174 non-null   float64
 10  new_sp_f1524  3194 non-null   float64
 11  new_sp_f2534  3200 non-null   float64
 12  new_sp_f3544  3199 non-null   float64
 13  new_sp_f4554  3204 non-null   float64
 14  new_sp_f5564  3195 non-null   float64
 15  new_sp_f65    3197 non-null   float64
 16  new_sn_m014   1045 non-null   float64
 17  new_sn_m1524  1030 non-null   float64
 18  new_sn_m2534  1022 non-null 

## Question 1. We need to know what's included in this data.
- How many countries are present?  
- What's the timespan of the data?  
- Does each country have a row for every year present?  
- Which countries are missing years?  
- How many rows have at least 1 non-null values across the columns 3-57?

In [133]:
count = df.country.unique().size
print("Number of countries in the dataset:", count)

Number of countries in the dataset? 219


In [32]:
# What's the timespan of data
a = df.year.min()
b = df.year.max()

print(f"We have data from {a} to {b}")

We have data from 1980 to 2013


In [34]:
#Does each country have a row for every year present?

result = []
for country in df.country.unique():
    cntry_df = df[df['country'] == country]
    if len(cntry_df.year.unique()) != len(df.year.unique()):
        result.append(country)
        
print("Below are the list of countries where there isn't a row for every year present:\n", result)

Below are the list of countries where there isn't a row for every year present:
 ['Bonaire, Saint Eustatius and Saba', 'Curacao', 'Montenegro', 'Netherlands Antilles', 'Serbia', 'Serbia & Montenegro', 'Sint Maarten (Dutch part)', 'South Sudan', 'Timor-Leste']


#### Ans. Based on the above output it is clear that not country has a row for every year that is present

In [42]:
# List of column names from 3-57
list_of_cols = df.columns.values.tolist()[2:58]

df[list_of_cols].dropna(thresh=1)

new_sp_m014  new_sp_m1524  new_sp_m2534  new_sp_m3544  new_sp_m4554  \
17            0.0          10.0           6.0           3.0           5.0   
18           30.0         129.0         128.0          90.0          89.0   
19            8.0          55.0          55.0          47.0          34.0   
20           52.0         228.0         183.0         149.0         129.0   
21          129.0         379.0         349.0         274.0         204.0   
...           ...           ...           ...           ...           ...   
7235        125.0         578.0           NaN        3471.0         681.0   
7236        150.0         710.0        2208.0        1682.0         761.0   
7237        152.0         784.0        2467.0        2071.0         780.0   
7238        120.0         783.0        2421.0        2086.0         796.0   
7239          NaN           NaN           NaN           NaN           NaN   

      new_sp_m5564  new_sp_m65  new_sp_f014  new_sp_f1524  new_sp_f2534  ...  \
17             2.0         0.0          5.0          38.0          36.0  ...   
18            64.0        41.0         45.0         350.0         419.0  ...   
19            21.0         8.0         25.0         139.0         160.0  ...   
20            94.0        80.0         93.0         414.0         565.0  ...   
21           139.0       103.0        146.0         799.0         888.0  ...   
...            ...         ...          ...           ...           ...  ...   
7235         293.0       192.0        180.0         873.0           NaN  ...   
7236         350.0       252.0        173.0         974.0        2185.0  ...   
7237         377.0       278.0        174.0        1084.0        2161.0  ...   
7238         360.0       271.0        173.0         939.0        2053.0  ...   
7239           NaN         NaN          NaN           NaN           NaN  ...   

      newrel_m4554  newrel_m5564  newrel_m65  newrel_f014  newrel_f1524  \
17             NaN           NaN         NaN          NaN           NaN   
18             NaN           NaN         NaN          NaN           NaN   
19             NaN           NaN         NaN          NaN           NaN   
20             NaN           NaN         NaN          NaN           NaN   
21             NaN           NaN         NaN          NaN           NaN   
...            ...           ...         ...          ...           ...   
7235           NaN           NaN         NaN          NaN           NaN   
7236           NaN           NaN         NaN          NaN           NaN   
7237           NaN           NaN         NaN          NaN           NaN   
7238           NaN           NaN         NaN          NaN           NaN   
7239        2349.0        1206.0      1208.0       1252.0        2069.0   

      newrel_f2534  newrel_f3544  newrel_f4554  newrel_f5564  newrel_f65  
17             NaN           NaN           NaN           NaN         NaN  
18             NaN           NaN           NaN           NaN         NaN  
19             NaN           NaN           NaN           NaN         NaN  
20             NaN           NaN           NaN           NaN         NaN  
21             NaN           NaN           NaN           NaN         NaN  
...            ...           ...           ...           ...         ...  
7235           NaN           NaN           NaN           NaN         NaN  
7236           NaN           NaN           NaN           NaN         NaN  
7237           NaN           NaN           NaN           NaN         NaN  
7238           NaN           NaN           NaN           NaN         NaN  
7239        4649.0        3526.0        1453.0         811.0       725.0  

[3484 rows x 56 columns]

#### Ans. 3484 rows have atleast 1 non-null values across the columns 3-57

## Question 2: Wide data isn't usually helpful.
Convert this to a long format with the following columns:  
1. country  
2. year  
3. diagnosis method: mapping should be: 

```
  {'newsp':'Pulmonary Smear (+)'
  ,'newsn':'Pulmonary Smear (-)'
  ,'newep':'Extrapulmonary'
  ,'newrel':'Relapse'}   
 ```

4. gender: male or female.   
5. age: lower_age - higher_age, e.g., 0-14  
6. number of cases  

__Note__: You can use `replace` (https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.replace.html)

In [98]:
# Dropping all rows where there no non-null values

new_df = df[list_of_cols].dropna(thresh=1)
new_df = pd.merge(df, new_df, how='inner', on=list_of_cols)
new_df = new_df.drop_duplicates()
new_df

country  year  new_sp_m014  new_sp_m1524  new_sp_m2534  \
0     Afghanistan  1997          0.0          10.0           6.0   
1     Afghanistan  1998         30.0         129.0         128.0   
2     Afghanistan  1999          8.0          55.0          55.0   
3     Afghanistan  2000         52.0         228.0         183.0   
4     Afghanistan  2001        129.0         379.0         349.0   
...           ...   ...          ...           ...           ...   
4379     Zimbabwe  2009        125.0         578.0           NaN   
4380     Zimbabwe  2010        150.0         710.0        2208.0   
4381     Zimbabwe  2011        152.0         784.0        2467.0   
4382     Zimbabwe  2012        120.0         783.0        2421.0   
4383     Zimbabwe  2013          NaN           NaN           NaN   

      new_sp_m3544  new_sp_m4554  new_sp_m5564  new_sp_m65  new_sp_f014  ...  \
0              3.0           5.0           2.0         0.0          5.0  ...   
1             90.0          89.0          64.0        41.0         45.0  ...   
2             47.0          34.0          21.0         8.0         25.0  ...   
3            149.0         129.0          94.0        80.0         93.0  ...   
4            274.0         204.0         139.0       103.0        146.0  ...   
...            ...           ...           ...         ...          ...  ...   
4379        3471.0         681.0         293.0       192.0        180.0  ...   
4380        1682.0         761.0         350.0       252.0        173.0  ...   
4381        2071.0         780.0         377.0       278.0        174.0  ...   
4382        2086.0         796.0         360.0       271.0        173.0  ...   
4383           NaN           NaN           NaN         NaN          NaN  ...   

      newrel_m4554  newrel_m5564  newrel_m65  newrel_f014  newrel_f1524  \
0              NaN           NaN         NaN          NaN           NaN   
1              NaN           NaN         NaN          NaN           NaN   
2              NaN           NaN         NaN          NaN           NaN   
3              NaN           NaN         NaN          NaN           NaN   
4              NaN           NaN         NaN          NaN           NaN   
...            ...           ...         ...          ...           ...   
4379           NaN           NaN         NaN          NaN           NaN   
4380           NaN           NaN         NaN          NaN           NaN   
4381           NaN           NaN         NaN          NaN           NaN   
4382           NaN           NaN         NaN          NaN           NaN   
4383        2349.0        1206.0      1208.0       1252.0        2069.0   

      newrel_f2534  newrel_f3544  newrel_f4554  newrel_f5564  newrel_f65  
0              NaN           NaN           NaN           NaN         NaN  
1              NaN           NaN           NaN           NaN         NaN  
2              NaN           NaN           NaN           NaN         NaN  
3              NaN           NaN           NaN           NaN         NaN  
4              NaN           NaN           NaN           NaN         NaN  
...            ...           ...           ...           ...         ...  
4379           NaN           NaN           NaN           NaN         NaN  
4380           NaN           NaN           NaN           NaN         NaN  
4381           NaN           NaN           NaN           NaN         NaN  
4382           NaN           NaN           NaN           NaN         NaN  
4383        4649.0        3526.0        1453.0         811.0       725.0  

[3484 rows x 58 columns]

In [131]:
dfLong = (
    new_df.melt(id_vars=['country', 'year'], var_name='diagnosis')
)

dfLong.head()

country  year    diagnosis  value
0  Afghanistan  1997  new_sp_m014    0.0
1  Afghanistan  1998  new_sp_m014   30.0
2  Afghanistan  1999  new_sp_m014    8.0
3  Afghanistan  2000  new_sp_m014   52.0
4  Afghanistan  2001  new_sp_m014  129.0

In [132]:
dfLong['gender'] = [x.replace('new_', 'new').split('_')[1] for x in dfLong['diagnosis']]
dfLong['diagnosis'] = [x.replace('new_', 'new').split('_')[0] for x in dfLong['diagnosis']]
dfLong['age'] = [x[1:2]+'-'+x[2:] if len(x[1:])==3 else x[1:3]+'-'+x[3:] for x in dfLong['gender']]
dfLong['age'] = [x[:-1] if x[-1]=='-' else x for x in dfLong['age']]
dfLong['gender'] = ['male' if x[0:1]=='m' else 'female' for x in dfLong['diagnosis']]
dfLong.rename(columns={"value": "no_of_cases"}, inplace=True)

dfLong

country  year diagnosis  no_of_cases  gender   age
0       Afghanistan  1997     newsp          0.0  female  0-14
1       Afghanistan  1998     newsp         30.0  female  0-14
2       Afghanistan  1999     newsp          8.0  female  0-14
3       Afghanistan  2000     newsp         52.0  female  0-14
4       Afghanistan  2001     newsp        129.0  female  0-14
...             ...   ...       ...          ...     ...   ...
195099     Zimbabwe  2009    newrel          NaN  female    65
195100     Zimbabwe  2010    newrel          NaN  female    65
195101     Zimbabwe  2011    newrel          NaN  female    65
195102     Zimbabwe  2012    newrel          NaN  female    65
195103     Zimbabwe  2013    newrel        725.0  female    65

[195104 rows x 6 columns]

## Question 3: Create a graph that shows when countries started to report TB cases.

In [134]:
dfLong.sort_values(by=['country', 'no_of_cases'], ascending=True)

country  year diagnosis  no_of_cases  gender    age
0       Afghanistan  1997     newsp          0.0  female   0-14
20904   Afghanistan  1997     newsp          0.0  female     65
41808   Afghanistan  1997     newsp          0.0  female  55-64
48786   Afghanistan  2007     newsn          0.0  female   0-14
52270   Afghanistan  2007     newsn          0.0  female  15-24
...             ...   ...       ...          ...     ...    ...
195098     Zimbabwe  2008    newrel          NaN  female     65
195099     Zimbabwe  2009    newrel          NaN  female     65
195100     Zimbabwe  2010    newrel          NaN  female     65
195101     Zimbabwe  2011    newrel          NaN  female     65
195102     Zimbabwe  2012    newrel          NaN  female     65

[195104 rows x 6 columns]